### PR TITLE
Sw improvement

### PR DESCRIPTION
**Pull request by mhadaily**
Hi @housseindjirdeh, 

I spent a bit time and try to achieve 100 out of 100 in lighthouse for each of modules. However, I could improve these score a bit. 

That would be great if you check and we can discuss to achieve even more on performance which is 88 right now. 

I think each commit should tell you what I have done, just wanna mention that I optimize images with Google Guetzli manually.

### CODE REVIEW

**Comment by housseindjirdeh** [comment]
This is amazing :heart_eyes:, can't thank you enough for this!

So happy to hear that you would like to help optimize this app further, absolutely let's definitely try more perf techniques  and see how much we can improve it.

Regarding `StyleExtHtmlWebpackPlugin`, does internalizing styles into `<style>` help with performance? (genuinely curious as I'm not sure). Other than that, this PR looks awesome, can you please also add yourself as a contributor to the README? :)

**Comment by mhadaily** [comment]
Well, .css file blocks rendering the rest of page, thus, having critical css  in <style> would help performance, although it's not significant. In our case, It improved only 0.5 second. However, it solved another problem which is  "block rendering" file while parsing the html. That's why performance has incremented by 4 to 88.  The better option would be to add only  critical css to head instead of whole content of css file.

I saw you haved tried Http2 push, if we can get that in place, so then we can have probably .css file and have it in http2 push. 

There are still some ways that we need to try to boost performance 💯 . I am more than happy if we can discuss about more perf techniques and get it to improve the page performance. 

Feel free to revert any commit that seems not appropriate to you. 👍

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -11,11 +11,11 @@ Even if you have little to no experience with Angular, I'll be more than happy t
...
 5. `ng init`
 6. Type `n` for each file to not overwrite any file changes
-7. Run `ng serve` and open `localhost:4200` in a browser
+7. Run `npm start` and open `localhost:4200` in a browser
 8. Work your magic
-9. Run `ng build --prod --aot` or `ng serve --prod --aot` to kick off a production build and make sure nothing is broken
+9. Run `npm run build` or `npm run static-serve` to kick off a production build and make sure nothing is broken
 10. To test service worker changes:
-  * `ng build --prod --aot` to kick off a fresh build and update the `dist/` directory
+  * `npm run build` to kick off a fresh build and update the `dist/` directory
```
Comment (housseindjirdeh) [inline CONTRIBUTING.md]: Thank you for updating this file as well :tada:

File: package.json [hunk 2]
```diff
@@ -10,31 +10,31 @@
     "precache": "sw-precache --verbose --config=sw-precache-config.js",
     "static-serve": "cd dist && live-server --port=4200 --host=localhost --entry-file=/index.html",
     "bundle-report": "webpack-bundle-analyzer dist/stats.json",
-    "build": "webpack",
+    "build": "rimraf dist && NODE_ENV=production webpack",
```
Comment (housseindjirdeh) [inline package.json]: :+1:

File: src/main.ts [hunk 3]
```diff
@@ -9,4 +9,12 @@ if (environment.production) {
   enableProdMode();
 }
 
-platformBrowserDynamic().bootstrapModule(AppModule);
+platformBrowserDynamic().bootstrapModule(AppModule)
```
Comment (housseindjirdeh) [inline src/main.ts]: Instead of relying on `window`, this makes sense!

File: src/manifest.json [hunk 4]
```diff
@@ -11,7 +11,12 @@
 			"src": "assets/icons/android-chrome-256x256.png",
 			"sizes": "256x256",
 			"type": "image\/png"
-		}
+		},
+    {
+      "src": "assets/icons/android-chrome-512x512.png",
```
Comment (housseindjirdeh) [inline src/manifest.json]: :+1:

File: webpack.config.js [hunk 5]
```diff
@@ -196,7 +197,7 @@ module.exports = {
         "manifest.json"
       ],
       "globOptions": {
-        "cwd": "/Users/houssein/Dev/angular2-hn/src",
+        "cwd": "./src",
```
Comment (housseindjirdeh) [inline webpack.config.js]: Cheers, that's a good catch!

File: src/index.html [hunk 6]
```diff
@@ -44,11 +44,16 @@
   <link rel="apple-touch-icon" sizes="120x120" href="assets/icons/apple-touch-icon-120x120.png">
   <link rel="apple-touch-icon" sizes="152x152" href="assets/icons/apple-touch-icon-152x152.png">
   <link rel="apple-touch-icon" sizes="180x180" href="assets/icons/apple-touch-icon-180x180.png">
+  <style>
+    #skip a {  position:absolute;  left:-10000px;  top:auto;  width:1px;  height:1px;  overflow:hidden;  }
+    #skip a:focus {  position:static;  width:auto;  height:auto;  }
+  </style>
 </head>
 <body>
+  <div id="skip"><a href="#content">skip to navigation</a></div>
```
Comment (housseindjirdeh) [inline src/index.html]: This is awesome thank you <3

**Comment by housseindjirdeh** [comment]
Okay that is awesome that we see some performance wins internalizing our styles with `<style>`. And you're right, running Lighthouse on latest I do see that it's render blocking. And everything else in the PR looks good!

Yep I've tried h2 push but didn't notice any particular wins. This was sometime ago so I'm more then interested to giving it a shot again. What I tried in the meantime was preloading our critical link tags (`<link rel=preload`> using `PreloadWebpackPlugin` but unfortunately didn't catch any significant wins there as well. However, I opened #67 to further explore h2 push!

Thanks again for such a good well-rounded PR. You've noticed areas of accessibility that needed improvement as well as performance hits. Can't thank you enough!

**Pull request merged**